posts: add notice about private broadcast IP leak in 31.0 (#1255)
bitcoin-core/bitcoincore.org: _posts/en/posts/2026-06-06-privatebroadcast-ip-leak.md

---
title: Private Broadcast May Reveal Sender IP Address in Bitcoin Core 31.0
name: privatebroadcast-ip-leak
id: en-privatebroadcast-ip-leak
lang: en
type: posts
layout: post

> excerpt: A bug in the -privatebroadcast feature, newly introduced in Bitcoin Core 31.0, may reveal the originator's IP address to the receiving peer under certain network conditions. Workarounds are available, and the issue is fixed in 31.1.

Reviewer: ```suggestion
excerpt: A bug in the -privatebroadcast feature, newly introduced in Bitcoin Core 31.0, may reveal the originator's IP address to the receiving peer under certain network conditions. Workarounds are available, and the issue is fixed in the upcoming 31.1.
```

---

We have become aware of a privacy bug in the `-privatebroadcast` feature, newly
introduced in Bitcoin Core 31.0, that may cause the originator's IP address to be
revealed to the receiving peer under certain network conditions. A fix is

> forthcoming and will be released as 31.1. Users of `-privatebroadcast` are

Reviewer 2: “as 31.1” seems odd to me.

```suggestion
revealed to the receiving peer under certain network conditions. A fix is
forthcoming and will be released with 31.1. Users of `-privatebroadcast` are
```

Author: Done.

advised to apply one of the workarounds below until 31.1 is released.

## Affected users

> This bug affects users where all the following are true:

Reviewer: Since disabling v2 transport is listed as a workaround, shouldn't the affected-users list also say that v2 transport must be enabled? Otherwise the list also includes users who have already applied the workarounds.

Reviewer 3: Yeah, I was also thinking to adding that `-v2trasport=1` must be used in order for the bug to possibly surface. I didn't mention it because I thought that is not needed since `-v2trasport=1` is the default. I am fine either way.

Author: Done.

> - The node is running Bitcoin Core 31.0 and `-privatebroadcast` is set.

Reviewer 2: I think it would make sense to mention v2transport here, too:

```suggestion
- The node is running Bitcoin Core 31.0, `-v2transport` is enabled (this is the default), and `-privatebroadcast` is set.
```

Author: Added a line about `-v2transport` not being disabled, but moved it to the end (since it's least likely to IMO to be the case where it's disabled).

  restriction excludes them, and no `-proxy=...` value applies to them.

## Impact

When private broadcast selects an IPv4 or IPv6 peer that advertises
support for v2 (BIP324) transport, the initial connection is routed
through the Tor proxy as expected. If the v2 handshake fails on that
connection, Bitcoin Core retries it as v1. The v1 retry is not routed
through the Tor proxy and instead connects directly to the peer over
IPv4 or IPv6, exposing the originator's IP address to the recipient.

Initial v1 connections (to peers that do not advertise v2) are
correctly routed through the Tor proxy and are not affected. The
bug is specific to the v1 reconnection that follows a failed v2
handshake. Connections to onion and I2P peers are also unaffected,
because they remain routed through their respective proxies on any v1
retry and therefore never expose a clearnet IP address.

This breaks the privacy guarantee stated in the 31.0 release notes:
"Their IP address (and thus geolocation) is never known to the recipients".

## Workarounds

Until they can upgrade to 31.1, users of `-privatebroadcast` should apply one
of the following:

> 1. **Disable the feature.** Set `-privatebroadcast=0`.

Reviewer: nit: would `noprivatebroadcast` also work? Probably not worth changing, just thinking out loud...
```suggestion
1. **Disable the feature.** Set `-noprivatebroadcast`.
```

Reviewer 4: IMO `-privatebroadcast=0` is clearer. The -no* prefixes are probably not well known to many users.

Reviewer 5: nononononoprivatebroadcast

Reviewer: ...=-1

2. **Disable v2 transport.** Set `-v2transport=0`. This causes all of the
   node's connections to use the unencrypted v1 protocol, which has the downside
   that it becomes easier to fingerprint and censor on clearnet.

3. **Route IPv4/IPv6 outbound through Tor.** Set
   `-proxy=127.0.0.1:9050` (replace `9050` with your Tor SOCKS
   port if different). This routes _all_ outbound IPv4/IPv6 P2P
   traffic through Tor exit nodes, which has the downside of making the node
   easier to Sybil attack.

## Credits

Credit to Eugene Siegel for discovering the bug.

> ## How this can happen

Reviewer: This is useful context for understanding when the leak can actually happen. I would move this before the workarounds, probably after the Impact section, instead of placing it after Credits.

Author: Done.

---
title: Private Broadcast May Reveal Sender IP Address in Bitcoin Core 31.0
name: privatebroadcast-ip-leak
id: en-privatebroadcast-ip-leak
lang: en
type: posts
layout: post

> excerpt: A bug in the -privatebroadcast feature of Bitcoin Core 31.0 may reveal the originator's IP address to the receiving peer under certain network conditions. Workarounds are available. A fix will be released as 31.1.

Reviewer 3: If this will stay on the web site after 31.1 is released, then the wording "will be released" will be incorrect. What about "Fixed in 31.1"?

Reviewer: I'd still say `and the issue is expected to be fixed in the upcoming 31.1 release.` but I'm fine with both.

Author: Taken, except without "is expected".

> excerpt: A bug in the -privatebroadcast feature of Bitcoin Core 31.0 may reveal the originator's IP address to the receiving peer under certain network conditions. Workarounds are available. A fix will be released as 31.1.

Reviewer: I'd mention that it's a new feature, just introduced here

## Impact

When private broadcast selects an IPv4 or IPv6 peer that advertises
support for v2 (BIP324) transport, the initial connection is routed
through the Tor proxy as expected. If the v2 handshake fails on that
connection, Bitcoin Core retries it as v1. The v1 retry is not routed
through the Tor proxy and instead connects directly to the peer over
IPv4 or IPv6, exposing the originator's IP address to the recipient.

Initial v1 connections (to peers that do not advertise v2) are
correctly routed through the Tor proxy and are not affected. The
bug is specific to the v1 reconnection that follows a failed v2
handshake. Connections to onion and I2P peers are also unaffected,
because they remain routed through their respective proxies on any v1
retry and therefore never expose a clearnet IP address.

This breaks the privacy guarantee stated in the 31.0 release notes:
"Their IP address (and thus geolocation) is never known to the recipients".

## Workarounds

Until 31.1 is released, users of `-privatebroadcast` should apply one
of the following:

1. **Disable the feature.** Set `-privatebroadcast=0`.

2. **Disable v2 transport.** Set `-v2transport=0`. This causes all of the

>    node's connections to use the unencrypted v1 protocol, which make it easier

Reviewer 3: ```suggestion
   node's connections to use the unencrypted v1 protocol, which has the downside that it becomes easier
```

   to fingerprint and censor on clearnet.

3. **Restrict the node to Tor (and optionally I2P).** Set
   `-onlynet=onion` (and `-onlynet=i2p` if I2P is configured).
   This disables IPv4 and IPv6 for the entire node. Note that
   restricting the node to a single network in this way can make it

>    significantly easier to Sybil attack.

Reviewer 3: To me 4. looks strictly better than this and so I would drop this. This requires the attacker to create many `.onion` nodes. 4. requires the attacker to create many `.onion` nodes plus the attacker to control part of the Tor network itself (the exit nodes).
